File: leetcode/solutions/0121.BestTimeToBuyAndSellStock/maxProfit.ts
export default function maxProfit(prices: number[]): number {
  let minPrice = Infinity;
  let maxProfit = 0;

  for (const currentPrice of prices) {
    minPrice = Math.min(minPrice, currentPrice);
    maxProfit = Math.max(maxProfit, currentPrice - minPrice);
  }

  return maxProfit;
}

File: leetcode/solutions/0121.BestTimeToBuyAndSellStock/maxProfit.go
package solution

func maxProfit(prices []int) int {
	if prices == nil || len(prices) == 0 {
		return 0
	}

	var maxProfit int
	minPrice := prices[0]

	for i := 1; i < len(prices); i++ {
		price := prices[i]

		if minPrice > price {
			minPrice = price
			continue
		}

		profit := price - minPrice

		if maxProfit < profit {
			maxProfit = profit
		}
	}

	return maxProfit
}
